### ISSUE TITLE
az containerapp update fails when run without permissions to the associated managed environment

### ISSUE THREAD

**Issue by maskati** [microsoft/azure-container-apps#530]
### This issue is a:
- [x] bug report -> please search issues before submitting
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)

### Issue description
The Azure CLI `az containerapp update` command (without `--yaml`) performs a full `PUT` ([Create or Update](https://learn.microsoft.com/en-us/rest/api/containerapps/container-apps/create-or-update)) of the container app, including updating `properties.managedEnvironmentId`. This causes issues in scenarios where the deployment credential does not have permissions to the container app associated managed environment.

### Steps to reproduce
Perform `az containerapp update --image` using a credential having only Contributor role (`Microsoft.App/containerApps/write` permission) to the Container App and no permissions to the currently associated managed environment.

**Expected behavior**
Since the `az containerapp update --image` command is only updating `properties.template.containers.image`, and is not joining the app to a new environment, this should not need permissions to the currently associated managed environment.

**Actual behavior**
The command fails with a `LinkedAuthorizationFailed` when the credential does not have `Microsoft.App/managedEnvironments/join/action` permission to the currently associated `properties.managedEnvironmentId`:

> (LinkedAuthorizationFailed) The client 'xxx' with object id 'xxx' has permission to perform action 'Microsoft.App/containerApps/write' on scope '/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.App/containerApps/xxx'; however, it does not have permission to perform action 'Microsoft.App/managedEnvironments/join/action' on the linked scope(s) '/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.App/managedEnvironments/xxx' or the linked scope(s) are invalid.

**Workaround**         
A workaround is to perform the image update using the [Update](https://learn.microsoft.com/en-us/rest/api/containerapps/container-apps/update) REST API directly, which does not touch `properties.managedEnvironmentId` and so works even when not having permissions to the managed environment:

`az rest --method patch --url "/subscriptions/{subscriptionId}/resourceGroups/xxx/providers/Microsoft.App/containerApps/xxx?api-version=2022-10-01" --body '{"properties":{"template":{"containers":[{"image":"myacr.azurecr.io/mymage:mytag","name":"mycontainer"}]}}}'`

**Comment by staterium** [comment]
I encountered the same error, although mine was from a Github action (azure/container-apps-deploy-action@v1). Here is how I fixed it:

- In Azure Portal, in the home page top searchbar, search for the client mentioned in the error message.
- You'll either get a result in the Azure AD section, or an item that says Continue searching in Azure AD
- When you've located the client in Azure AD (an Enterprise Application), you'll see the Object ID is equal to the Client ID in the error message. Copy the Name of the Application (the topmost value in the Overview screen).
- Navigate to your subscription, then to Access control (IAM).
- Click on Add, then Add role assignment.
- On the Role tab, select Priviledged administrator roles, then select Contributor. Click Next.
- On the Members tab, click Select members, and then paste the value of the Enterprise Application you copied earlier.
- Click Select, Close, then Next, and then Review + assign.

After that your container app should deploy.

**Comment by tadzjibov** [comment]
> I encountered the same error, although mine was from a Github action (azure/container-apps-deploy-action@v1). Here is how I fixed it:
> 
> * In Azure Portal, in the home page top searchbar, search for the client mentioned in the error message.
> * You'll either get a result in the Azure AD section, or an item that says Continue searching in Azure AD
> * When you've located the client in Azure AD (an Enterprise Application), you'll see the Object ID is equal to the Client ID in the error message. Copy the Name of the Application (the topmost value in the Overview screen).
> * Navigate to your subscription, then to Access control (IAM).
> * Click on Add, then Add role assignment.
> * On the Role tab, select Priviledged administrator roles, then select Contributor. Click Next.
> * On the Members tab, click Select members, and then paste the value of the Enterprise Application you copied earlier.
> * Click Select, Close, then Next, and then Review + assign.
> 
> After that your container app should deploy.

Thanks a lot! It worked for me

**Comment by maskati** [comment]
The suggested change does work, but do not address the original issue:

> in scenarios where the deployment credential does not have permissions to the container app associated managed environment

Deployment should work also in cases when such permission does not exist. Ideally the Azure CLI `az containerapp update --image` should only update `properties.template.containers.image` using the [Update](https://learn.microsoft.com/en-us/rest/api/containerapps/container-apps/update) API (HTTP PATCH).

**Comment by bgrainger** [comment]
If you don't want to grant the full Contributor role, there's a list of permissions for a custom role here https://github.com/microsoft/azure-container-apps/issues/35#issuecomment-975770598 (see also my comment below that) that should allow `az containerapp update` to work.

**Comment by maskati** [comment]
@bgrainger that is correct, although i still feel this is a bug in the CLI tooling, since technically you don't need any permissions to the environment in order to update a container app already associated with an environment. This might be the case for example when you have workload profiles with limited capacities, and you don't want the app deployment identity to be able to provision new workloads to the environment.

**Comment by bgrainger** [comment]
Agreed; I was posting it as a possible workaround for anyone else blocked by this issue.

FWIW, I tried the HTTP `PATCH` route that you suggested, but found that  `{"containers":[{"image":"myacr.azurecr.io/mymage:mytag","name":"mycontainer"}]}` destroyed any existing environment variables from the last revision. My assumption was that JSON Merge Patch would leave environment variables as-is unless `"env": null` was explicitly in the JSON, but it doesn't work that way, making the REST `PATCH` API pretty useless (for updating a container image) IMO. (Quite possibly the same is true for `probes` and `volumeMounts` but I wasn't using those so I didn't end up testing them.)

**Comment by mastoj** [comment]
I have the same issue when trying to use pulumi to create an application in a pre-existing environment. My user has the contributor role for both the application and environment, so it is very strange.
